clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -2421,6 +2413,38 @@ public void testGetOffsetsFencedLeaderEpoch() {
         assertEquals(0L, metadata.timeToNextUpdate(time.milliseconds()));
     }
 
+    @Test
+    public void testGetOffsetByTimeWithPartitionsRetryCouldTriggerMetadataUpdate() {
+        buildFetcher();
+        List<Errors> retriableErrors = Arrays.asList(Errors.NOT_LEADER_FOR_PARTITION,
+            Errors.REPLICA_NOT_AVAILABLE, Errors.KAFKA_STORAGE_ERROR, Errors.OFFSET_NOT_AVAILABLE,
+            Errors.LEADER_NOT_AVAILABLE, Errors.FENCED_LEADER_EPOCH, Errors.UNKNOWN_LEADER_EPOCH);
+
+        for (Errors retriableError : retriableErrors) {
+            subscriptions.assignFromUser(singleton(tp1));
+            client.updateMetadata(initialUpdateResponse);
+
+            Node originalLeader = metadata.fetch().leaderFor(tp1);
+
+            final int updatedNodeSize = 3;
+            MetadataResponse updatedMetadata = TestUtils.metadataUpdateWith("dummy", 3, Collections.emptyMap(), singletonMap(topicName, 4), tp -> 3);
+
+            client.prepareMetadataUpdate(updatedMetadata);
+            client.prepareResponseFrom(listOffsetResponse(tp1, retriableError, ListOffsetRequest.LATEST_TIMESTAMP, -1L), originalLeader);
+
+            final long timestamp = 1L;
+            Node newLeader = new Node(1, "localhost", 1970);
+            assertNotEquals(originalLeader, newLeader);
+
+            client.prepareResponseFrom(listOffsetResponse(tp1, Errors.NONE, timestamp, 5L), newLeader);
+            Map<TopicPartition, OffsetAndTimestamp> offsetAndTimestampMap =
+                fetcher.offsetsForTimes(Collections.singletonMap(tp1, timestamp), time.timer(Integer.MAX_VALUE));
+
+            assertEquals(Collections.singletonMap(tp1, new OffsetAndTimestamp(5L, timestamp)), offsetAndTimestampMap);
+            assertEquals(updatedNodeSize, metadata.fetch().nodes().size());
```

Comment:
Not really sure this has value if the test case expects the leader change correctly.

Reply:
Removed